server/src/main/java/org/elasticsearch/injection/Injector.java

```diff
@@ -0,0 +1,310 @@
+            logger.trace("Specs: {}", result.values().stream().map(Object::toString).collect(joining("\n\t", "\n\t", "")));
+        }
+        return result;
+    }
+
+    private static MethodHandleSpec methodHandleSpecFor(Class<?> c) {
+        Constructor<?> constructor = getSuitableConstructorIfAny(c);
+        if (constructor == null) {
+            throw new IllegalStateException("No suitable constructor for " + c);
+        }
+
+        MethodHandle ctorHandle;
+        try {
+            ctorHandle = lookup().unreflectConstructor(constructor);
+        } catch (IllegalAccessException e) {
+            throw new IllegalStateException(e);
+        }
+
+        List<ParameterSpec> parameters = Stream.of(constructor.getParameters()).map(ParameterSpec::from).toList();
+
+        return new MethodHandleSpec(c, ctorHandle, parameters);
+    }
+
+    private static void assertSeedMapIsValid(Map<Class<?>, InjectionSpec> seed) {
+        seed.forEach(
+            (c, s) -> { assert s.requestedType().equals(c) : "Spec must be associated with its requestedType, not " + c + ": " + s; }
+        );
+    }
+
+    /**
+     * For the classes we've been explicitly asked to inject,
+     * pretend there's some massive method taking all of them as parameters
+     */
+    private static ParameterSpec syntheticParameterSpec(Class<?> c) {
+        return new ParameterSpec("synthetic_" + c.getSimpleName(), c, c);
+    }
+
+    private static Constructor<?> getSuitableConstructorIfAny(Class<?> type) {
+        var constructors = Stream.of(type.getConstructors()).filter(not(Constructor::isSynthetic)).toList();
+        if (constructors.size() == 1) {
+            return constructors.get(0);
+        }
+        var injectConstructors = constructors.stream().filter(c -> c.isAnnotationPresent(Inject.class)).toList();
+        if (injectConstructors.size() == 1) {
+            return injectConstructors.get(0);
+        }
+        logger.trace("No suitable constructor for {}", type);
+        return null;
+    }
+
+    private static void registerSpec(InjectionSpec spec, Map<Class<?>, InjectionSpec> specsByClass) {
+        Class<?> requestedType = spec.requestedType();
+        var existing = specsByClass.put(requestedType, spec);
+        if (existing == null || existing.equals(spec)) {
+            logger.trace("Register spec: {}", spec);
+        } else {
+            throw new IllegalStateException("Ambiguous specifications for " + requestedType + ": " + existing + " and " + spec);
+        }
+    }
+
+    private List<InjectionStep> injectionPlan(Set<Class<?>> requiredClasses, Map<Class<?>, InjectionSpec> specsByClass) {
+        logger.trace("Constructing instantiation plan");
+        Set<Class<?>> allParameterTypes = new HashSet<>();
+        specsByClass.values().forEach(spec -> {
+            if (spec instanceof MethodHandleSpec m) {
+                m.parameters().stream().map(ParameterSpec::injectableType).forEachOrdered(allParameterTypes::add);
```

Comment:
Does it need to be ordered here if we're putting into a `HashSet`?

Reply:
`Stream.forEach` does not guarantee a "happens before" relation between the actions, so it's generally unsafe to use it for multiple updates to a data structure that isn't, say, a `ConcurrentHashMap`. The javadocs say:

> If the action accesses shared state, it is responsible for providing the required synchronization.

However, that's a bit academic because we know this particular stream is sequential rather than parallel. If we added `.parallel()`, my code would keep working while `forEach` would not; but let's face it, we're not going to do that.

I tend to default to `forEachOrdered` unless I know the action is safe to do in parallel, but I admit that's not the common convention. I can change it to `forEach` if you find this justification unconvincing. 😄

Reply:
I wish we had been given three variants:
- `forEachOrdered` guarantees both encounter order and happens-before
- `forEach` guarantees happens-before
- `forEachParallel` guarantees neither

Reply:
@thecoop - by the javadocs, the question of whether to use `forEachOrdered` isn't only about whether the order matters, but also about whether the action being executed is prepared to handle concurrency.

Anyway I don't feel strongly about this and I can remove the `Ordered` if the team prefers it.